content/tutorials/parallelization/SIMWE_parallelization.qmd

```diff
@@ -0,0 +1,804 @@
+basin_map.show()
+```
+
+![](SIMWE_images/basin_030401010402_both.webp)
+
+# Parallelized hydrologic and sediment modeling
+
+To efficiently model erosion and deposition across a large number of HUC12 subwatersheds,
+we will use the workflow we just ran and create a
+script that uses Python's `multiprocessing` module to parallelize the workflow.
+Each subwatershed is processed independently in its own environment, which allows computations to run concurrently without interference.
+
+Each subwatershed needs to set different mask and computational region, however normally, those settings are global, and so for different mask and region for each parallel process, we will use mask and region context managers.
+
+* Masking is handled using `MaskManager`, a [context manager for setting and managing raster mask](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.MaskManager),
+making it possible to have custom mask for the current process. This feature is available only since GRASS 8.5.
+
+    ```python
+    with gs.MaskManager(mask_name=f"basin_{huc12}"):
+        # Run actual computation with active mask.
+        gs.run_command(...)
+    ```
+
+* Computational region is handled using `RegionManager`, a [context manager for setting and managing computational region](https://grass.osgeo.org/grass-devel/manuals/libpython/grass.script.html#grass.script.RegionManager),
+making it possible to have custom region for the current process. This feature is available only since GRASS 8.5.
+
+    ```python
+    with gs.RegionManager(vector=f"basin_{huc12}"):
+        # Run actual computation in the specified region.
+        gs.run_command(...)
+    ```
+
+## Putting it all together
+
+We will put all the pieces together into a script that runs each subwatershed in parallel.
+The resulting statistics for each subwatershed are collected and stored in a JSON file for further analysis.
+
+```{python}
+%%writefile script.py
+import os
+import json
+from multiprocessing import Pool, cpu_count
+from tqdm import tqdm
+
+import grass.script as gs
+from grass.tools import Tools
+
+
+def compute(huc12):
+    simulation_time = 180
+    # set overwrite to True to rerun the workflow
+    tools = Tools(overwrite=True, quiet=True)
+    tools.v_extract(
+        input="river_basins",
+        output=f"basin_{huc12}",
+        where=f"huc12 == '{huc12}'",
+        flags="t",
+    )
+    # Set the computational region to match the basin
+    with gs.RegionManager(vector=f"basin_{huc12}", raster="ned"):
+        tools.v_to_rast(
+            input=f"basin_{huc12}",
+            output=f"basin_{huc12}",
+            use="val",
+        )
+        with gs.MaskManager(mask_name=f"basin_{huc12}"):
+            # Run actual computation with active mask.
+            tools.r_proj(
+                project="nlcd",
+                mapset="PERMANENT",
+                input="nlcd",
+                output=f"nlcd_{huc12}",
+            )
```

Comment:
Will this use the mask in any way? Maybe just put it to the region section to set the expectations right.